### PR TITLE
[RB-31] Sign Up

### PR DESCRIPTION
**Pull request by EvgeniiRychkov**
## Description
Create Sign Up page

## Changes
Create Sign Up page, create Input component

## Related issues
RB-31

## Screenshots
Screenshot 2024-11-20 at 12 01 39 AM

### CODE REVIEW

File: src/api/DBRequests.jsx [hunk 1]
```diff
@@ -15,3 +15,22 @@ export const login = async ({ email, password }) => {
...
+        email,
+        password,
+        firstName,
+        lastName,
+      }
+    );
+    return { data, status };
+  } catch (error) {
+    const errorMessage =
+      error?.response?.data?.error || 'An unexpected error occurred';
```
Comment (in43sh) [inline src/api/DBRequests.jsx]: You could extract `An unexpected error occurred` to a const variable. Small change but might be worth doing it considering that we will have more requests in the future

File: src/api/DBRequests.jsx [hunk 2]
```diff
@@ -15,3 +15,22 @@ export const login = async ({ email, password }) => {
     throw new Error(errorMessage);
   }
 };
+
+export const register = async ({ email, password, firstName, lastName }) => {
```
Comment (in43sh) [inline src/api/DBRequests.jsx]: You could combine these two functions into one:
```js
import axios from 'axios';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL;

const handleApiRequest = async (url, payload) => {
  try {
    const { data, status } = await axios.post(`${API_BASE_URL}${url}`, payload);
    return { data, status };
  } catch (error) {
    const errorMessage =
      error?.response?.data?.message || 'An unexpected error occurred';
    throw new Error(errorMessage);
  }
};

export const login = (credentials) =>
  handleApiRequest('/api/v1/login', credentials);

export const register = (userData) =>
  handleApiRequest('/api/v1/register', userData);

```

File: src/components/Form/Input.jsx [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+const Input = ({ id, name, type, value, onChange, children, error }) => {
```
Comment (in43sh) [inline src/components/Form/Input.jsx]: You can remove `return` here:
```js
const Input = ({ id, name, type, value, onChange, children, error }) => (
  <div className="mb-3 flex w-full flex-col gap-1">
    <label htmlFor={id}>{children}</label>
    <input
      id={id}
      name={name}
      type={type}
      value={value}
      onChange={onChange}
      className={'w-full rounded border-2 border-gray p-2'}
    />
    {error && <p className="text-red">{error}</p>}
  </div>
);
```

File: src/hooks/UseSignUpForm.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    email: '',
+    password: '',
+    firstName: '',
+    lastName: '',
+  });
+  const [error, setError] = useState({});
+  const { setUserSession } = useAuth();
+  const navigate = useNavigate();
+
+  const handleChange = ({ target: { name, value } }) => {
```
Comment (in43sh) [inline src/hooks/UseSignUpForm.jsx]: Very minor but if you want, you could remove curly braces:
```js
const handleChange = ({ target: { name, value } }) =>
    setForm((prevForm) => ({ ...prevForm, [name]: value }));
```

File: src/hooks/UseSignUpForm.jsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    if (!form.lastName) errors.lastName = 'Last Name is required';
+    return errors;
+  };
+
+  const handleSubmit = async (e) => {
+    e.preventDefault();
+
+    const validationErrors = validateForm();
+    setError(validationErrors);
+    if (Object.keys(validationErrors).length > 0) {
```
Comment (in43sh) [inline src/hooks/UseSignUpForm.jsx]: You could replace it with smaller code piece:
```js
if (Object.keys(validationErrors).length) return;
```

File: src/hooks/UseSignUpForm.jsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+      return;
+    }
+
+    try {
+      const result = await register(form);
```
Comment (in43sh) [inline src/hooks/UseSignUpForm.jsx]: You can use destructuring here as you've done before by doing:
```js
const { data, status } = await register(form);
```

File: src/hooks/UseSignUpForm.jsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import { useState } from 'react';
+import { useNavigate } from 'react-router-dom';
+import { register } from '../api/DBRequests';
+import { useAuth } from '../context/AuthProvider';
+
+const UseSignUpForm = () => {
```
Comment (in43sh) [inline src/hooks/UseSignUpForm.jsx]: As this is a React custom hook, I think you should rename both the function and the filename to `useSignUpForm`

File: src/pages/auth/SignIn.jsx [hunk 8]
```diff
@@ -19,33 +20,29 @@ const SignIn = () => {
...
-              onChange={handleChange}
-              className="w-full rounded border-2 border-gray p-2"
-            />
-            {error.email && <p className="text-red">{error.email}</p>}
-          </div>
+          <Input
+            id="email"
+            name="email"
+            type="email"
+            onChange={handleChange}
```
Comment (in43sh) [inline src/pages/auth/SignIn.jsx]: I think `value` should go before `onChange`. Here and in all other cases

File: src/pages/auth/SignIn.jsx [hunk 9]
```diff
@@ -19,33 +20,29 @@ const SignIn = () => {
...
+            error={error.password}
+          >
+            Password
+          </Input>
 
-          <p className="mb-7">
+          <p>
             <Link
               to="/password/reset"
               className="flex w-full flex-col underline"
```
Comment (in43sh) [inline src/pages/auth/SignIn.jsx]: Do you really need all these classes for the Link? Shouldn't it be better to only have `underline`?

**Review by in43sh — CHANGES_REQUESTED** [review]

**Comment by in43sh** [comment]
Looks awesome!

**Review by brandon91warren — APPROVED** [review]

**Review by in43sh — APPROVED** [review]

**Pull request merged**